### ISSUE TITLE
Enable GMPXX automatically

### ISSUE THREAD

**Issue by mglisse** [CGAL/cgal#4266]
## Issue Details

With old CGAL, I could configure with `-DWITH_GMPXX`, which defined some macro in `compiler_config.h` and enabled GMPXX for all programs using CGAL. It was already sad that I had to specify `-DWITH_GMPXX` once when configuring, since CMake could have looked for it and enabled it if available. Now with header-only, I actually need to do that every time I use CGAL. Would it be acceptable to enable GMPXX automatically if it is detected? One way would be to use `find_package(GMPXX QUIET)` in `CGAL_SetupGMP.cmake`. Another one would rely on `__has_include`, which is a bit more dangerous because it does not affect the linker.

I understand that some users might want tight control of exactly what is used, but I believe most just want CGAL to sort its implementation details itself, which means enabling GMPXX. Or the documentation on how to use CGAL should explain very clearly that users are supposed to enable everything they have available on their machine.

## Environment

* Operating system (Windows/Mac/Linux, 32/64 bits): Linux
* Compiler: g++-9
* Release or debug mode: Release
* CGAL version: 5.0-beta1
* Other libraries versions if used (Eigen, TBB, etc.): GMP-6.1.2

**Comment by lrineau** [comment]
I do not know how to react to that. Some say the auto-detection of optional dependencies is evil, and should be avoided, because it does not give reproducible builds.

  https://youtu.be/sBP17HQAQjk?t=650

And I understand the reasoning of it. At least we should provide *another* CMake option *and* another preprocessor macro to disable the auto-detection.

**Comment by mglisse** [comment]
Yes, I know some don't like it, that was the beginning of my second paragraph. I would argue for making GMPXX a required dependency (in the same sense as GMP) if it wasn't so hard to use with MSVC.
Now some builds will have GMPXX, some won't. Meaning that some will run faster, some slower (well, with a recent enough boost and boost.multiprecision as a fallback, maybe not by much, but let's ignore that, and detecting the version of boost is just as evil). Not detecting it means all would run slow, not really that great.
Maybe when I run cmake to configure CGAL at the beginning, if I pass WITH_GMPXX, it could take note in CGALConfig.cmake (or a file it includes) and enable GMPXX by default for users. This way for instance a linux distribution could easily decide to make GMPXX a hard dependency.
I didn't think that long about the various options, mostly it feels frustrating that we went through the trouble of making CGAL work well with GMPXX and so few people benefit.

**Comment by zhfsu** [comment]
Hi, mglisse, I agree with you. I like auto-detection and agrees that GMPXX be a required dependency of CGAL. May I ask you two questions here? 1) is GMPXX a C++ version of GMP? 2) Is there any precompiled binaries of GMPXX downloadable for Windows? CGAL has a precompiled version of GMP & MPFR for Windows shipped with it, it would be nice to have GMPXX lib too.

**Comment by mglisse** [comment]
> 1) is GMPXX a C++ version of GMP?

It is a C++ interface for GMP, which comes as part of GMP.

> 2) Is there any precompiled binaries of GMPXX downloadable for Windows?

No, and that's probably not possible, because of the way visual studio works. It would need to be one precompiled binary per version of visual studio (and per set of options like release/debug?), too much work. Many applications can use GMPXX without linking with the library, but we cannot guarantee that. MPIR, with its GMP-compatibility mode, might (or not) be usable there.

**Comment by zhfsu** [comment]
Hi, mglisse, Thank you for the the reply. 
In addition, I have a good news for Windows GMPXX users. The GMP pre-compiled binary shipped with CGAL for Windows is working perfectly with GMPXX interface. Also, in view of the fact that the CGAL's pre-compiled GMP binary has nothing to do with VS version, we don't have to build it for each VS version. Yeah~. Some notes when using CGAL's pre-compiled GMP binary and GMPXX interface for Windows: 1) Turn off the warning/error of "C4146: unary minus operator applied to unsigned type, result still unsigned" by Project Property -> C/C++ -> All Options -> Additional Options -> /w34146 (or /wd4146 to disable it completely depending on your needs) 2) Since the version of CGAL's pre-compiled GMP binary is 5.0.1, while gmpxx.h I could find is from GMP's website which is 6.1.2, VS2015 will complain that it can not recognize two new functions -- mpq_cmp_z & mpf_cmp_z -- when compiling gmpxx.h. So we need to add the following macro definitions into the code before #include <gmpxx.h>, which are borrowed from gmp-h.in:

> #define mpq_cmp_z __gmpq_cmp_z
> __GMP_DECLSPEC int mpq_cmp_z(mpq_srcptr, mpz_srcptr) __GMP_ATTRIBUTE_PURE;
> #define mpf_cmp_z __gmpf_cmp_z
> __GMP_DECLSPEC int mpf_cmp_z(mpf_srcptr, mpz_srcptr) __GMP_NOTHROW __GMP_ATTRIBUTE_PURE;

An example of GMPXX code embedded in Eigen3 which I have tested to be working in VS2015 on Windows 10 is as follows:
```
#include <iostream>
#include <Eigen/Core>
#include <gmp.h>

#define mpq_cmp_z __gmpq_cmp_z
__GMP_DECLSPEC int mpq_cmp_z(mpq_srcptr, mpz_srcptr) __GMP_ATTRIBUTE_PURE;
#define mpf_cmp_z __gmpf_cmp_z
__GMP_DECLSPEC int mpf_cmp_z(mpf_srcptr, mpz_srcptr) __GMP_NOTHROW __GMP_ATTRIBUTE_PURE;

#include <gmpxx.h>

std::ostream& operator<<(std::ostream& os, mpz_class data) {
	return os << data.get_d();
}

typedef Eigen::Matrix<mpz_class, Eigen::Dynamic, Eigen::Dynamic> MatrixXz;

int main() {
	MatrixXz A(10, 10), B(10, 10), C(10, 10);
	A.fill(111);
	B.fill(222);
	C = 2 * A + B;
	std::cout << C << std::endl;
}

```
So, that's it. Enjoy GMPXX, Windows programmers!

**Comment by mglisse** [comment]
@zhfsu No, it does not work "perfectly". As I wrote
> Many applications can use GMPXX without linking with the library, but we cannot guarantee that

In particular, if you use C++ I/O, you will be in trouble. Using numeric_limits in debug mode is also likely to fail.
Your declaration of `mpq_cmp_z` is also insufficient in case anyone compares an integer with a rational, it just delays the error. You can download older versions of GMP from `https://gmplib.org/download/gmp/`.

**Comment by zhfsu** [comment]
Thank you for letting me know the download link of older versions of GMP. I downloaded version 5.0.1 and tried to use the gmpxx.h therein so that I don't have to mix two versions of GMP. The problems I encountered is that when compiling `default_digits10_impl<T,false,false>::run()` in `Eigen\src\Core\NumTraits.h`, VS2015 complains that `log10` does not support the argument type `__gmp_expr<mpz_t, mpz_t>` That's not surprising because that type is not considered in the design of the function. To solve this issue, I used class template specialization and put it immediately after `#include <gmpxx.h>`:

```
namespace Eigen {
namespace internal {
	template<>
	struct default_digits10_impl<__gmp_expr<mpz_t, mpz_t>, false, false>
	{
		static int run() {
			return 0;
		}
	};
}
}

```
Now the complain disappeared and my code runs correctly.

As for C++ I/O, the following code indeed reports an error:

> 	mpz_class a, b, c;
> 
> 	a = 1234;
> 	b = "-5678";
>      std::cout << "absolute value is " << abs((a + b) / 2) << "\n";

But if we simply apply type conversion as follows, no problem would occur:
```
std::cout << "absolute value is " << mpz_class(abs((a + b) / 2)) << "\n";
```

When it comes to `mpq_cmp_z`, I agree that it is insufficient if anyone compares an integer with a rational, but I think that's the business of GMP's designer, and since that's the new function introduced in the latest version, I don't have to worry about it now because I'm using v5.0.1 for the moment (thank mglisse again for telling me the download link. I was thinking of using gmpxx.h of version 5.0.1, but I just couldn't find it).

Maybe it is not "perfect", but I can fix the problem bit by bit and my application finally works (correctly). That's all I want and that's already perfect for me :-). Thank everyone who helps.
